storey/flow.py

```diff
@@ -175,33 +175,41 @@ async def _run_loop(self):
             for path in self._paths:
                 async with aiofiles.open(path, mode='r') as f:
                     header = None
+                    reverse_header = None
                     if self._with_header:
                         line = await f.readline()
-                        if self._build_dict:
-                            header = next(csv.reader([line]))
+                        header = next(csv.reader([line]))
+                        reverse_header = {}
+                        for i in range(len(header)):
+                            reverse_header[header[i]] = i
                     async for line in f:
                         parsed_line = next(csv.reader([line]))
                         element = parsed_line
                         key = None
                         if header:
                             if len(parsed_line) != len(header):
                                 raise ValueError(f'CSV line with {len(parsed_line)} fields did not match header with {len(header)} fields')
-                            element = {}
-                            for i in range(len(parsed_line)):
-                                if self._build_dict:
+                            if self._build_dict:
+                                element = {}
+                                for i in range(len(parsed_line)):
                                     element[header[i]] = parsed_line[i]
-                                else:
+                            else:
+                                for i in range(len(parsed_line)):
                                     element[i] = parsed_line[i]
                         if self._key_field:
-                            key = element[self._key_field]
-                            del element[self._key_field]
+                            key_field = self._key_field
+                            if reverse_header and isinstance(key_field, str):
+                                key_field = reverse_header[key_field]
+                            key = parsed_line[key_field]
```

Comment:
If `self._key_field` is a column name (aka string), and the user didn't specify `with_header` it's best to return a nice error rather than fail here. probably need some input validation in the ctor